src/routes/version_creation.rs

```diff
@@ -0,0 +1,239 @@
+    pub mod_id: Option<ModId>,
+    pub file_parts: Vec<String>,
+    pub version_number: String,
+    pub version_title: String,
+    pub version_body: String,
+    pub dependencies: Vec<VersionId>,
+    pub game_versions: Vec<GameVersion>,
+    pub release_channel: VersionType,
+    pub loaders: Vec<ModLoader>,
+}
+
+#[post("api/v1/version")]
+pub async fn version_create(
+    payload: Multipart,
+    client: Data<PgPool>,
+    file_host: Data<std::sync::Arc<dyn FileHost + Send + Sync>>,
+) -> Result<HttpResponse, CreateError> {
+    let mut transaction = client.begin().await?;
+    let mut uploaded_files = Vec::new();
+
+    let result = version_create_inner(
+        payload,
+        &mut transaction,
+        &***file_host,
+        &mut uploaded_files,
+    )
+    .await;
+
+    if result.is_err() {
+        let undo_result = super::mod_creation::undo_uploads(&***file_host, &uploaded_files).await;
+        let rollback_result = transaction.rollback().await;
+
+        if let Err(e) = undo_result {
+            return Err(e);
+        }
+        if let Err(e) = rollback_result {
+            return Err(e.into());
+        }
+    } else {
+        transaction.commit().await?;
+    }
+
+    result
+}
+
+async fn version_create_inner(
+    mut payload: Multipart,
+    transaction: &mut sqlx::Transaction<'_, sqlx::Postgres>,
+    file_host: &dyn FileHost,
+    uploaded_files: &mut Vec<UploadedFile>,
+) -> Result<HttpResponse, CreateError> {
+    //TODO: Check if mod exists, check if version number is unique
+    let cdn_url = dotenv::var("CDN_URL")?;
+
+    let mut initial_version_data = None;
+    let mut version_builder = None;
+
+    while let Some(item) = payload.next().await {
+        let mut field: Field = item.map_err(CreateError::MultipartError)?;
+        let content_disposition = field.content_disposition().ok_or_else(|| {
+            CreateError::MissingValueError("Missing content disposition".to_string())
+        })?;
+        let name = content_disposition
+            .get_name()
+            .ok_or_else(|| CreateError::MissingValueError("Missing content name".to_string()))?;
+
+        if name == "data" {
+            let mut data = Vec::new();
+            while let Some(chunk) = field.next().await {
+                data.extend_from_slice(&chunk.map_err(CreateError::MultipartError)?);
+            }
+
+            let version_create_data: InitialVersionData = serde_json::from_slice(&data)?;
+            initial_version_data = Some(version_create_data.clone());
+
+            let mod_id: ModId = version_create_data.mod_id.ok_or_else(|| {
+                CreateError::InvalidInput("Mod id is required for version data".to_string())
+            })?;
+
+            let results = sqlx::query!(
+                "SELECT EXISTS(SELECT 1 FROM mods WHERE id=$1)",
+                mod_id.0 as i64
+            )
+            .fetch_one(&mut *transaction)
+            .await?;
+
+            if !results.exists.unwrap_or(true) {
```

Comment:
This should be `.unwrap_or(false)`, since the unwrapped value is immediately inverted.  If the query fails (I don't think it can?) it should act like it didn't find any mods for that id.